### PR TITLE
Implement hard drive smart error monitoring

### PR DESCRIPTION
**Pull request by bugagashenkj**

### CODE REVIEW

File: lib/health.js [hunk 1]
```diff
@@ -170,6 +170,42 @@ const cmdExec = (cmd, callback) => {
   });
 };
 
+const getSmartStatusOfDrive = (driveName, callback) => {
+  const cmd = `smartctl -H /dev/${driveName} | grep overall`;
```
Comment (nechaido) [inline lib/health.js]: I'd prefer to use `rg`, `ag` if either is installed.

@belochub WDYT?

File: lib/health.js [hunk 2]
```diff
@@ -170,6 +170,42 @@ const cmdExec = (cmd, callback) => {
...
+  api.cp.exec(cmd, (error, stdout) => {
+    if (error) callback(error);
+
+    const status = stdout.toString().split(':').pop().trim();
+    if (status === 'PASSED') {
```
Comment (nechaido) [inline lib/health.js]:
```suggestion
    callback(null, status === 'PASSED')
```

**Comment by bugagashenkj** [comment]
@tshemsedinov Should we support disk monitoring for Windows?

File: lib/health.js [hunk 3]
```diff
@@ -170,6 +170,38 @@ const cmdExec = (cmd, callback) => {
   });
 };
 
+const getSmartStatusOfDrive = (driveName, callback) => {
+  const cmd = `smartctl -H /dev/${driveName}`;
+  api.cp.exec(cmd, (error, stdout) => {
+    if (error) callback(error);
```
Comment (belochub) [inline lib/health.js]: You don't return after calling the callback here, thus it is possible that the callback will be called twice here.
Also, I think it'd be better to not ignore `stderr`, and print it in this branch/pass it to the callback.

File: lib/health.js [hunk 4]
```diff
@@ -170,6 +170,38 @@ const cmdExec = (cmd, callback) => {
...
+    callback(null, status === 'PASSED');
+  });
+};
+
+
+health.getSmartStatus = callback => {
+  const cmd = 'lsblk -d -J';
+  const drivesStatus = {};
+  api.cp.exec(cmd, (error, stdout) => {
+    if (error) callback(error);
```
Comment (belochub) [inline lib/health.js]: Same thing here.

File: lib/health.js [hunk 5]
```diff
@@ -170,6 +170,38 @@ const cmdExec = (cmd, callback) => {
...
+
+    const funcs = JSON.parse(stdout).blockdevices.map(drive =>
+      next => getSmartStatusOfDrive(drive.name, (error, status) => {
+        if (error) next(error);
```
Comment (belochub) [inline lib/health.js]: Ditto.

File: lib/health.js [hunk 6]
```diff
@@ -170,6 +170,53 @@ const cmdExec = (cmd, callback) => {
...
+const getSmartStatusOfDrive = (driveName, callback) => {
+  const cmd = `smartctl -H /dev/${driveName}`;
+  api.cp.exec(cmd, (error, stdout, stderr) => {
+    if (error) {
+      callback(error);
+      return;
+    }
+
+    if (stderr) {
+      callback(stderr);
```
Comment (belochub) [inline lib/health.js]: `stderr` is a string here, thus you should either log it here or create an error based on it, not pass it to the callback

File: lib/health.js [hunk 7]
```diff
@@ -170,6 +170,53 @@ const cmdExec = (cmd, callback) => {
...
+    if (error) {
+      impress.log.error(stderr);
+      callback(error);
+      return;
+    }
+
+    const funcs = JSON.parse(stdout).blockdevices.map(drive =>
+      next => getSmartStatusOfDrive(drive.name, (error, status) => {
+        if (error) {
+          impress.log.error(stderr);
```
Comment (belochub) [inline lib/health.js]: Why do you log the `stderr` obtained when running `lsblk -d -J` when you get an error from `getSmarStatusOfDrive()`?
It is now possible for it to be logged more than once.

File: lib/health.js [hunk 8]
```diff
@@ -170,6 +170,49 @@ const cmdExec = (cmd, callback) => {
...
+        if (error) {
+          next(error);
+          return;
+        }
+        drivesStatus[drive.name] = status;
+        next(null);
+      })
+    );
+
+    api.metasync.sequential(funcs, (error) => {
```
Comment (tshemsedinov) [inline lib/health.js]:
```suggestion
    api.metasync.sequential(funcs, error => {
```

**Comment by tshemsedinov** [comment]
Landed in 4050d2e

**Pull request closed without merging**